File: python/201312-3-最大的矩形-2-递增栈-优化.py
n = int(input())
heights = [int(x) for x in input().split()]
stack = [-1]
res = 0
heights.append(-1)

for idx, val in enumerate(heights):
    while val < heights[stack[-1]]:  #如果小于栈顶元素则谈栈
        h = heights[stack.pop()]
        res = max(res, h * (idx - stack[-1] - 1))   #每次弹栈后，都计算面积
    stack.append(idx)
print(res)

File: c++/201312-3-最大的矩形-2-递增栈-优化.cpp
#include <iostream>
#include <algorithm>
#include <stack>

using namespace std;

int heights[10005];
stack<int> st;

int main() {
    int n;
    cin >> n;
    for (int i = 0; i < n; i++) {
        cin >> heights[i];
    }
    heights[n] = -1; //追加一个小于任何数的 min辅助
    st.push(n); //栈中压入 min辅助 的索引
    int res = 0;
    for (int i = 0; i <= n; i++) {
        while (heights[i] < heights[st.top()]) {
            int h = heights[st.top()];
            st.pop();
            res = max(res, h * (i - (st.top() != n ? st.top() : -1) - 1));
        }
        st.push(i);
    }
    cout << res << endl;

    return 0;
}